esp-hal/src/uart/uhci.rs

```diff
@@ -0,0 +1,525 @@
+        {
+            unsafe {
+                self.channel
+                    .rx
+                    .prepare_transfer(self.uhci.dma_peripheral(), &mut rx_buffer)?
+            };
+
+            self.channel.rx.start_transfer()?;
+
+            Ok(UhciDmaRxTransfer::new(self, rx_buffer))
+        }
+    }
+}
+
+impl<'d> Uhci<'d, Blocking> {
+    /// todo
+    pub fn new(
+        uart: Uart<'d, Blocking>,
+        uhci: peripherals::UHCI0<'static>,
+        channel: impl DmaChannelFor<AnyUhci<'d>>,
+    ) -> Self {
+        let channel = Channel::new(channel.degrade());
+        channel.runtime_ensure_compatible(&uhci);
+
+        let uhci = Uhci {
+            uart,
+            uhci: uhci.into(),
+            channel,
+        };
+
+        uhci.init();
+        uhci
+    }
+
+    /// Create a new instance in [crate::Async] mode.
+    pub fn into_async(self) -> Uhci<'d, Async> {
+        Uhci {
+            uart: self.uart.into_async(),
+            uhci: self.uhci,
+            channel: self.channel.into_async(),
+        }
+    }
+}
+
+impl<'d> Uhci<'d, Async> {
+    /// Create a new instance in [crate::Blocking] mode.
+    pub fn into_blocking(self) -> Uhci<'d, Blocking> {
+        Uhci {
+            uart: self.uart.into_blocking(),
+            uhci: self.uhci,
+            channel: self.channel.into_blocking(),
+        }
+    }
+}
+
+// Based on SpiDmaTransfer
+/// A structure representing a DMA transfer for UHCI/UART.
+///
+/// This structure holds references to the UHCI instance, DMA buffers, and
+/// transfer status.
+pub struct UhciDmaTxTransfer<'d, Dm, Buf>
+where
+    Dm: DriverMode,
+    Buf: DmaTxBuffer,
+{
+    uhci: ManuallyDrop<Uhci<'d, Dm>>,
+    dma_buf: ManuallyDrop<Buf::View>,
+}
+
+impl<'d, Buf: DmaTxBuffer, Dm: DriverMode> UhciDmaTxTransfer<'d, Dm, Buf> {
+    fn new(uhci: Uhci<'d, Dm>, dma_buf: Buf) -> Self {
+        Self {
+            uhci: ManuallyDrop::new(uhci),
+            dma_buf: ManuallyDrop::new(dma_buf.into_view()),
+        }
+    }
+
+    /// Returns true when [Self::wait] will not block.
+    pub fn is_done(&self) -> bool {
+        self.uhci.channel.tx.is_done()
+    }
+
+    /// Cancels the DMA transfer.
+    pub fn cancel(mut self) -> (Uhci<'d, Dm>, Buf) {
+        self.uhci.channel.tx.stop_transfer();
+
+        let retval = unsafe {
+            (
+                ManuallyDrop::take(&mut self.uhci),
+                DmaTxBuffer::from_view(ManuallyDrop::take(&mut self.dma_buf)),
+            )
+        };
+        core::mem::forget(self);
+        retval
+    }
+}
+
+impl<'d, Buf: DmaTxBuffer> UhciDmaTxTransfer<'d, Blocking, Buf> {
+    /// Waits for the DMA transfer to complete.
+    ///
+    /// This method blocks until the transfer is finished and returns the
+    /// `Uhci` instance and the associated buffer.
+    pub fn wait(mut self) -> Result<(Uhci<'d, Blocking>, Buf::View), Error> {
```

Comment:
In the event of an error, the peripheral and the buffer should be returned too. See how `Camera` and `I8080` do it.